metals/src/main/scala/scala/meta/internal/metals/SyntheticDecorationsProvider.scala

```diff
@@ -0,0 +1,200 @@
+package scala.meta.internal.metals
+
+import scala.annotation.nowarn
+import scala.collection.mutable
+import scala.{meta => m}
+
+import scala.meta.internal.decorations._
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.parsing.Trees
+import scala.meta.internal.pc.DecorationKind
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.SyntheticDecoration
+import scala.meta.tokens.{Token => T}
+
+import org.eclipse.{lsp4j => l}
+
+final class SyntheticDecorationsProvider(
+    params: m.pc.VirtualFileParams,
+    trees: Trees,
+    userConfig: () => UserConfiguration,
+) {
+
+  val path: AbsolutePath = params.uri().toAbsolutePath
+
+  lazy val (withoutTypes, methodPositions) = declarationsWithoutTypes()
+
+  def provide(
+      synthteticDecorations: List[SyntheticDecoration]
+  ): List[DecorationOptions] = {
+    val grouped = groupByRange(synthteticDecorations)
+    makeDecorations(grouped)
+
+  }
+
+  def makeDecoration(
+      pos: l.Position,
+      label: String,
+  ): DecorationOptions = {
+    val lspRange = new l.Range(pos, pos)
+    new DecorationOptions(
+      lspRange,
+      renderOptions = ThemableDecorationInstanceRenderOptions(
+        after = ThemableDecorationAttachmentRenderOptions(
+          label,
+          color = "grey",
+          fontStyle = "italic",
+          opacity = 0.7,
+        )
+      ),
+    )
+  }
+
+  private def groupByRange(
+      decorations: List[SyntheticDecoration]
+  ): List[List[SyntheticDecoration]] =
+    decorations
+      .groupBy(d => (d.range(), d.kind()))
+      .toList
+      .sortWith { case (((r1, k1), _), ((r2, k2), _)) =>
+        if (r1.lt(r2)) true
+        else if (r2.lt(r1)) false
+        else k1 < k2
+      }
+      .map(_._2)
+
+  @nowarn
+  private def makeDecorations(
+      grouped: List[List[SyntheticDecoration]]
+  ) = {
+    val result = mutable.ListBuffer.empty[DecorationOptions]
+    grouped
+      .map { case decorations @ (d :: _) =>
+        if (d.kind == DecorationKind.InferredType) {
+          val decoration = decorations.head
+          (decoration.range, decoration.label(), d.kind)
+        } else {
+          val labels = decorations.map(_.label()).reverse.mkString(", ")
+          (d.range, labels, d.kind)
+        }
+      }
+      .foreach { case (range, label, kind) =>
+        kind match {
+          case DecorationKind.ImplicitParameter =>
+            result += makeDecoration(
+              range.getStart(),
+              "(" + label + ")",
```

Comment:
Why not put it into label right away in the compiler?

Reply:
We could for this case, but for type parameters and implicit parameters, we want to zip multiple decorations into one here (so instead of `A[T][U][V]` we get `A[T,U,V]`). I though it's better to do it here, so we reuse logic between scala 2 and 3. I can move it to `Compilers` instead if its better?

Reply:
Should we always zip those? Wouldn't that be a problem in some cases such as interleaving clauses? I think it might be easier to do in the compiler, since we have lists of parameters etc.